adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/X509AuthHandler.java

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.adapter.http;
+
+import java.net.HttpURLConnection;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Objects;
+
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.security.auth.x500.X500Principal;
+
+import org.eclipse.hono.client.HonoClient;
+import org.eclipse.hono.service.auth.device.HonoAuthHandler;
+import org.eclipse.hono.service.auth.device.HonoClientBasedAuthProvider;
+import org.eclipse.hono.util.CredentialsConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.impl.HttpStatusException;
+
+
+/**
+ * An authentication handler for client certificate based authentication.
+ *
+ */
+public class X509AuthHandler extends HonoAuthHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(X509AuthHandler.class);
+    private static final HttpStatusException UNAUTHORIZED = new HttpStatusException(HttpURLConnection.HTTP_UNAUTHORIZED);
+    private final HonoClient tenantServiceClient;
+
+    /**
+     * Creates a new handler for an authentication provider and a
+     * Tenant service client.
+     * 
+     * @param authProvider The authentication provider to use for verifying
+     *              the device identity.
+     * @param tenantServiceClient The client to use for determining the tenant
+     *              that the device belongs to.
+     */
+    public X509AuthHandler(final HonoClientBasedAuthProvider authProvider, final HonoClient tenantServiceClient) {
+        super(authProvider);
+        this.tenantServiceClient = Objects.requireNonNull(tenantServiceClient);
+    }
+
+    @Override
+    public void parseCredentials(final RoutingContext context, final Handler<AsyncResult<JsonObject>> handler) {
+
+        final Future<JsonObject> result = Future.future();
+        result.setHandler(handler);
+        if (context.request().isSSL()) {
+            try {
+                final Certificate[] clientChain = context.request().sslSession().getPeerCertificates();
```

Comment:
My understanding is that right now only the first element of the certificate chain is being processed. This is fine but might not be enough for future scenarios. So I think it would make sense to extract this handling into another method, keep it as is for the moment. This would also make easier to read IMHO and could be separately tested. Which might be good for a security related topic.

Reply:
you mean something like
```
protected JsonObject getCredentialsFromCertificateChain(final Certificate[] certChain) {
...
}
```

Reply:
I was thinking about something like this (just roughly wrote it down, not checked):

~~~java
public void parseCredentials(final RoutingContext context, final Handler<AsyncResult<JsonObject>> handler) { {
       final Future<JsonObject> result = Future.future();
        result.setHandler(handler);

        if (!context.request().isSSL()) {
           result.fail(UNAUTHORIZED);
        }

        try {
            final Certificate[] clientChain = context.request().sslSession().getPeerCertificates();
            return processCertificate (clientChain);
        } catch (SSLPeerUnverifiedException e) {
            // client certificate has not been validated
            result.fail(UNAUTHORIZED);
        }
}

protected Future<JsonObject> processCertificate(RoutingContext context, Certificate[] clientChain) {
  if (!(clientChain[0] instanceof X509Certificate)) {
    LOG.info("cannot authenticate device using unsupported certificate type [{}]",
            clientChain[0].getClass().getName());
    return Future.failed(UNAUTHORIZED);
  }

  final X509Certificate clientCert = (X509Certificate) clientChain[0];
  return tenantServiceClient.getOrCreateTenantClient().compose(tenantClient ->
      tenantClient.get(clientCert.getIssuerX500Principal())
  ).map(tenant -> {
      return fromCertificate(tenant, clientCert);
  });

}

protected JsonObject fromCertificate (Tenant tenant, X509Certificate clientCert) {
      final String subjectDn = clientCert.getSubjectX500Principal().getName(X500Principal.RFC2253);
      LOG.debug("authenticating device of tenant [{}] using X509 certificate [subject DN: {}]",
              tenant.getTenantId(), subjectDn);

      return new JsonObject()
              .put(CredentialsConstants.FIELD_PAYLOAD_SUBJECT_DN, subjectDn)
              .put(CredentialsConstants.FIELD_PAYLOAD_TENANT_ID, tenant.getTenantId());
}
~~~

This would split apart the different aspects of the validation. Allow for extending/overriding those different aspects. To me this is also more readable, but that is just my personal preference.

Reply:
LGTM but I do not see why parseCredentials or processCertificate should be declared non-final. In the end, the specific implementation of these methods is what defines the Hono X509AuthHandler.